File: sorting/quickSort.js
// const pivotHelper = function (arr, start, end) {
//   let pivot = start;
//   let index = start;

//   for (let i = start + 1; i <= end; i++) {
//     if (arr[i] <= arr[pivot]) {
//       index++;
//       [arr[index], arr[i]] = [arr[i], arr[index]];
//     }
//   }
//   [arr[index], arr[pivot]] = [arr[pivot], arr[index]];

//   return index;
// };

// const quickSort = function (arr, start, end) {
//   if (start < end) {
//     let index = pivotHelper(arr, start, end);
//     quickSort(arr, start, index - 1);
//     quickSort(arr, index + 1, end);
//   }
//   return arr;
// };

// console.log(quickSort([10, 4, 2, 16, 0, 12, 1, 24], 0, 7));

const quickSort = function (arr, start, end) {
  if (start >= end) return;

  let pivot = end;
  let index = start - 1;

  for (let i = start; i <= end; i++) {
    if (arr[i] <= arr[pivot]) {
      index++;
      [arr[index], arr[i]] = [arr[i], arr[index]];
    }
  }
  quickSort(arr, start, index - 1);
  quickSort(arr, index + 1, end);
};

const arr = [1, 6, 2, 3, 15, 2, 3, 23, 43, 5];
quickSort(arr, 0, arr.length - 1);
console.log(arr);

File: sorting/quickSort.cpp
# include <iostream>
using namespace std;

void quickSort(int arr[], int start, int end) {
  if (start >= end) {
    return;
  }

  int pivot = end;
  int index = start - 1;

  for (int i = start; i <= end; i++) {
    if (arr[i] <= arr[pivot]) {
      index++;
      int temp = arr[i];
      arr[i] = arr[index];
      arr[index] = temp;
    }
  }
  quickSort(arr, start, index - 1);
  quickSort(arr, index + 1, end);
}

int main() {
  int arr[] = {1, 12, 15, 6, 2238, 9, 110, 12, 135};
	int size = sizeof(arr) / sizeof(arr[0]);

	for (int i = 0; i < size; i++) {
    	cout << arr[i] << " ";
	}
  cout << endl;

  quickSort(arr, 0, size - 1);

  for (int i = 0; i < size; i++) {
    cout << arr[i] << " ";
  }
  cout << endl;
  return 0;
}
